File: apps/ios/QuranMeali/Models/Verse.swift
//
//  Verse.swift
//  QuranMeali
//

import Foundation

struct Translation: Decodable, Hashable {
    let sourceCode: String
    let sourceName: String
    let languageCode: String
    let author: String
    let text: String
}

struct VerseSummary: Decodable, Identifiable, Hashable {
    let surahNumber: Int
    let verseNumber: Int
    let globalVerseNumber: Int
    let juzNumber: Int
    let pageNumber: Int
    let arabicText: String
    let translations: [Translation]

    var id: Int { globalVerseNumber }

    /// First available translation (the default source surfaces first
    /// thanks to backend ordering).
    var primaryTranslation: Translation? { translations.first }
}

struct VerseSearchHit: Decodable, Identifiable, Hashable {
    let surahNumber: Int
    let surahNameTurkish: String
    let verseNumber: Int
    let globalVerseNumber: Int
    let arabicText: String
    let translationSourceCode: String
    let translationText: String

    var id: String { "\(globalVerseNumber)-\(translationSourceCode)" }
}

struct PagedResult<T: Decodable & Hashable>: Decodable, Hashable {
    let items: [T]
    let totalCount: Int64
    let page: Int
    let pageSize: Int
    let totalPages: Int
    let hasNextPage: Bool
    let hasPreviousPage: Bool
}

File: apps/android/app/src/main/java/com/bbozkurtcagri/quranmeali/model/Verse.kt
package com.bbozkurtcagri.quranmeali.model

import com.squareup.moshi.JsonClass

@JsonClass(generateAdapter = false)
data class Translation(
    val sourceCode: String,
    val sourceName: String,
    val languageCode: String,
    val author: String,
    val text: String
)

@JsonClass(generateAdapter = false)
data class VerseSummary(
    val surahNumber: Int,
    val verseNumber: Int,
    val globalVerseNumber: Int,
    val juzNumber: Int,
    val pageNumber: Int,
    val arabicText: String,
    val translations: List<Translation>
) {
    // Backend ilk sırada default kaynağı koyuyor; UI'da bu birinci stiline güvenir.
    val primaryTranslation: Translation? get() = translations.firstOrNull()
}

@JsonClass(generateAdapter = false)
data class VerseSearchHit(
    val surahNumber: Int,
    val surahNameTurkish: String,
    val verseNumber: Int,
    val globalVerseNumber: Int,
    val arabicText: String,
    val translationSourceCode: String,
    val translationText: String
)

@JsonClass(generateAdapter = false)
data class PagedResult<T>(
    val items: List<T>,
    val totalCount: Long,
    val page: Int,
    val pageSize: Int,
    val totalPages: Int,
    val hasNextPage: Boolean,
    val hasPreviousPage: Boolean
)
